### DISCUSSION TITLE
Guidance for saving and restoring flow definitions in a backwards compatible manner

### DISCUSSION THREAD

**Discussion by yogiduzit** [xyflow/xyflow#4712] [General]
Consider a use-case where users can create multiple flows (with custom nodes), save them to a database and fetch (read: restore) them. The [Save and Restore](https://reactflow.dev/examples/interaction/save-and-restore) documentation recommends using the `toObject` method on a flow and save its output.

- In case of backwards-incompatible releases, will XYFlow team provide a backwards-compatible deserializer method that can convert (older) flow JSON into current flow definitions?
- Is the XYFlow team planning any backwards-incompatible changes akin to v10 migration where `elements` array was separated into `nodes` and `edges` array?
- Does the XYFlow team recommend storing flow definitions using the `toObject` method to databases? Is there a way to just store a minimal definition that doesn't involve coordinates, UI properties? Consumers can provide a minimal definition to XYFlow and it can auto-layout (maybe using dagre?) and assume reasonable defaults for UI properties.

**Comment by bcakmakoglu** [comment]
Can't really respond to the first two questions as that would need a roadmap that, as far as I am aware, doesn't exist (yet?), so who knows.

> * Does the XYFlow team recommend storing flow definitions using the toObject method to databases? Is there a way to just store a minimal definition that doesn't involve coordinates, UI properties? Consumers can provide a minimal definition to XYFlow and it can auto-layout (maybe using dagre?) and assume reasonable defaults for UI properties.

You are always free to remove properties from nodes/edges that are returned from `toObject`.
Nothing is stopping you from modifying nodes/edges in a way and into a shape that *you want to save*.

Auto-layouting will not be a core feature (and I doubt it ever will) since XYFlow (React and Svelte) are not concerned with solving use-cases *for you* but instead provide you the tools to solve these *by yourself*.
If you want to auto-layout, go ahead and do it but the core library itself will not attempt this for you.

**Reply by SimonSimCity** [reply]
I'm quite late to the discussion, but I want to rephrase something about the initial question, which might help others form a decision on this topic:

I am aware of that I (as a developer using this library in my codebase) can practically do what I want with the data returned to `toObject`, and that my custom data [should live within the `node.data` property](https://github.com/xyflow/xyflow/discussions/5022). Having this in mind:

Is there anything else I should keep in mind, or am I safe to assume that [diagrams saved on an older version can be restored](https://reactflow.dev/examples/interaction/save-and-restore) in any newer version? Of course, the future is impossible to predict, but any strategic advice from the side of the maintainers of this library would be nice.

One promise on your side could be never to change the structure in an incompatible way, but that might be impossible to keep at some state, so you might promise to - if the structure changes in an incompatible way - to have migrations available that you will apply while loading data from an older version. You might not want to have this around indefinitely and so on.

All this would give us developers using this library an even more solid ground for when to express changes and in what speed we would have to adopt to it.
